src/transaction_impl.cpp

```diff
@@ -147,13 +147,53 @@ void Transaction::Impl::Write(const std::string_view key,
   write_set_.emplace_back(std::move(sp));
 }
 
+void Transaction::Impl::Delete(const std::string_view key) {
+  if (IsAborted()) return;
+  EnsureCurrentTable();
+  bool deleted = current_table_->GetPrimaryIndex().Delete(key);
+  if (!deleted) {
+    Abort();
+    return;
+  }
+
+  bool is_rmf = false;
+  for (auto& snapshot : read_set_) {
+    if (snapshot.key == key &&
+        snapshot.table_name == current_table_->GetTableName()) {
+      is_rmf = true;
+      snapshot.is_read_modify_write = true;
+      break;
+    }
+  }
+
+  for (auto& snapshot : write_set_) {
+    if (snapshot.key != key ||
+        snapshot.table_name != current_table_->GetTableName())
+      continue;
+    snapshot.data_item_copy.Reset(nullptr, 0);
+    if (is_rmf) snapshot.is_read_modify_write = true;
+    return;
+  }
+
+  auto* index_leaf = current_table_->GetPrimaryIndex().GetOrInsert(key);
+
+  concurrency_control_->Write(key, nullptr, 0, index_leaf);
+  Snapshot sp(key, nullptr, 0, index_leaf, current_table_->GetTableName());
+  if (is_rmf) sp.is_read_modify_write = true;
+  write_set_.emplace_back(std::move(sp));
```

Comment:
Doesn't this code flow seem quite strange?

```
# 1. index->range_index->Delete()
bool deleted = current_table_->GetPrimaryIndex().Delete(key);

# 2. index->Insert()
auto* index_leaf = current_table_->GetPrimaryIndex().GetOrInsert(key);

# 3. index->Write(null)
concurrency_control_->Write(key, nullptr, 0, index_leaf);
```
I think we should write an outline of how Delete is implemented in the comments.

Reply:
I'd like to discuss this a bit. How should we implement `Delete` ?

1. **Logical deletion**. Set the DataItem's `IsInitialized` flag to false. `Scan()` skips items where `IsInitialized == false`. Items remain in both range_index and point_index.
2. **Physical deletion**. Free the DataItem and remove the item from both range_index and point_index.
3. **Hybrid**. Physically delete the key from range_index and do `Write(null)` to the DataItem in the point_index.

Your choice was option 3. It's not bad, but could we make it a physical delete (option 2)? That should be simpler.
In principle, since indexes are pluggable in LineairDB, this should be generic code usable even when adopting indexes other than precision_locking_index. From that perspective, the hybrid approach assumes that the index is split into point_index and range_index.

Reply:
I came up with some simple code of your solution (option 3). How about this?

```suggestion
  // Delete() consists of two deletions: removal from the index (physical)
  //   and initialization of the data item (logical).
  // The reason for this design is that we consider Delete() as
  //   a combination of two writes: a write to the index and a write to the data item.

  bool deleted = current_table_->GetPrimaryIndex().Delete(key);
  this->Write(key, nullptr, 0);
```